### PR TITLE
Optimize bsf, bsr, register cmpxchg, pcmpistri

### PR DESCRIPTION
**Pull request by alyssarosenzweig**
Just some random little things from looking at Secondary.json

### CODE REVIEW

File: unittests/ASM/TwoByte/0F_BC.asm [hunk 1]
```diff
@@ -41,6 +38,8 @@ bsf r12w, word  [rdx + 8 * 1]
 bsf r11d, dword [rdx + 8 * 1]
 bsf r10,  qword [rdx + 8 * 1]
 
+; bsf of zero has an undefined result in x86, so we don't check these results.
```
Comment (Sonicadvance1) [inline unittests/ASM/TwoByte/0F_BC.asm:41]: This isn't entirely true for bsf and bsr, we provide the stricter guarantees from AMD's platform. Intel is the only one with undefined (Although they also don't modify the destination)
> If the second operand contains 0, the instruction sets ZF to 1 and does not change the contents of the destination register

Comment (alyssarosenzweig) [reply unittests/ASM/TwoByte/0F_BC.asm:41]: I see. Is there a good reason to provide the AMD semantics when the documented behaviour is undefined? This undefined behaviour lets us save 2 instructions in the case that flags are discarded (1 instruction when flags are read).

Comment (Sonicadvance1) [reply unittests/ASM/TwoByte/0F_BC.asm:41]: I think you're misunderstanding. The documented behaviour is both, the quote I posted was from AMD's documentation, the quote you posted was from Intel's. Both are correct.

I believe I implemented the "correct" AMD documented behaviour once I saw an application relying on it years ago. Which very well could be the same reason why Intel hardware has the same behaviour but their documentation claims undefined.

Comment (alyssarosenzweig) [reply unittests/ASM/TwoByte/0F_BC.asm:41]: I see. I'll add a comment.

Comment (alyssarosenzweig) [reply unittests/ASM/TwoByte/0F_BC.asm]: Updated to keep the old semantics (but implemented more efficiently), dropped the unit test change, and added a comment.

**Review by Sonicadvance1 — APPROVED** [review]
nice nice nice

**Pull request merged**